File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py
import sys
from collections import deque, defaultdict

input = sys.stdin.readline
N, M, K, X = map(int, input().split())
graph = defaultdict(list)

for i in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)

q = deque([(X, 0)])
visit = [False] * (N + 1)
visit[X] = True
check = False
result = []

while q:
    cur_graph, dist = q.popleft()

    if dist == K:
        result.append(cur_graph)
        check = True

    for g in graph[cur_graph]:
        if not visit[g]:
            visit[g] = True
            q.append((g, dist + 1))

if not check:
    print(-1)
else:
    result.sort()
    for g in result:
        print(g)

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.cc
#include <bits/stdc++.h>
using namespace std;

int main()
{
  ios::sync_with_stdio(false);
  cin.tie(NULL);

  int N, M, K, X;

  cin >> N >> M >> K >> X;
  vector<vector<int>> graph(N + 1);
  vector<int> dist(N + 1, -1);

  for (int i = 0; i < M; i++)
  {
    int from, to;
    cin >> from >> to;
    graph[from].push_back(to);
  }

  deque<int> q;
  dist[X] = 0;
  q.push_back(X);

  while (!q.empty())
  {
    int cur_graph = q.front();
    q.pop_front();

    for (int next : graph[cur_graph])
    {
      if (dist[next] == -1)
      {
        dist[next] = dist[cur_graph] + 1;
        q.push_back(next);
      }
    }
  }

  bool check = false;

  for (int i = 1; i <= N; i++)
  {
    if (dist[i] == K)
    {
      check = true;
      cout << i << endl;
    }
  }

  if (!check)
  {
    cout << -1 << endl;
  }

  return 0;
}